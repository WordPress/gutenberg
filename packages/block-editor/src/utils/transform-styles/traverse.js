/**
 * External dependencies
 */
import traverse from 'traverse';

/**
 * Internal dependencies
 */
import { parse, stringify } from './ast';

function traverseCSS( css, callback ) {
	try {
		const parsed = parse( css );

		const updated = traverse.map( parsed, function( node ) {
			if ( ! node ) {
				return node;
			}
			const updatedNode = callback( node );
			return this.update( updatedNode );
		} );

		return stringify( updated );
	} catch ( err ) {
		// eslint-disable-next-line no-console
		console.warn( 'Error while traversing the CSS: ' + err );

		return null;
	}
}

export default traverseCSS;
