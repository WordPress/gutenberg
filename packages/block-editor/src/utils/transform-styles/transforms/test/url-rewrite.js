/**
 * Internal dependencies
 */
import traverse from '../../traverse';
import rewrite from '../url-rewrite';

describe( 'URL rewrite', () => {
	it( 'should replace relative paths', () => {
		const callback = rewrite( 'http://wp-site.local/themes/gut/css/' );
		const input = `h1 { background: url(images/test.png); }`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );

	it( 'should replace complex relative paths', () => {
		const callback = rewrite( 'http://wp-site.local/themes/gut/css/' );
		const input = `h1 { background: url(../images/test.png); }`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );

	it( 'should not replace absolute paths', () => {
		const callback = rewrite( 'http://wp-site.local/themes/gut/css/' );
		const input = `h1 { background: url(/images/test.png); }`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );

	it( 'should not replace remote paths', () => {
		const callback = rewrite( 'http://wp-site.local/themes/gut/css/' );
		const input = `h1 { background: url(http://wp.org/images/test.png); }`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );
} );
