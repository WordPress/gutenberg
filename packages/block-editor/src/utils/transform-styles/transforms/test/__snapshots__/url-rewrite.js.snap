// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`URL rewrite should not replace absolute paths 1`] = `
"h1 {
background: url(/images/test.png);
}"
`;

exports[`URL rewrite should not replace remote paths 1`] = `
"h1 {
background: url(http://wp.org/images/test.png);
}"
`;

exports[`URL rewrite should replace complex relative paths 1`] = `
"h1 {
background: url(http://wp-site.local/themes/gut/images/test.png);
}"
`;

exports[`URL rewrite should replace relative paths 1`] = `
"h1 {
background: url(http://wp-site.local/themes/gut/css/images/test.png);
}"
`;
